### PR TITLE
Added camera support on WalletConnect app 

### LINKED ISSUE #257: Add Camera support on WalletConnect app
## Overview

Having camera support on the wallet connect app would enable us to use tablets as a separate signing device

## Goals

enable tablet users to connect through walletconnect directly instead of typing/pasting the walletconnect uri

## Requirements

telegram-cloud-photo-size-1-5086951942919399975-y

a simple [o] button inside the typing area would be ideal

### PR DESCRIPTION
**Pull request by DaniSomoza**
## What it solves
Resolves #257

## How this PR fixes it

based on this code https://github.com/cozmo/jsQR/blob/master/docs/index.html

## How to test it

1. Go to [CowSwap](https://cowswap.exchange/#/swap) in a different device.
2. Click on connect wallet with Walletconnect > QR code
3. Go to wallet connect safe app in [this safe-react PR](https://github.com/gnosis/safe-react/pull/3310) ( Because Safe App iframe needs a special policy to be able to load the webcam).
4. Click on Load QR code icon in the input field.
5. Scan the QR Code from step 2.

## Screenshots

Captura de pantalla 2022-01-17 a las 10 46 16
Captura de pantalla 2022-01-17 a las 10 32 37
Captura de pantalla 2022-01-14 a las 15 28 12
Captura de pantalla 2022-01-17 a las 10 30 37
Captura de pantalla 2022-01-17 a las 10 35 54

### CODE REVIEW

File: apps/wallet-connect/src/App.tsx [hunk 1]
```diff
@@ -3,106 +3,33 @@ import jsQr from 'jsqr';
...
+import Dialog from '@material-ui/core/Dialog';
+import InputAdornment from '@material-ui/core/InputAdornment';
+import IconButton from '@material-ui/core/IconButton';
 
 import { blobToImageData } from './utils/images';
 import { Wrapper } from './components/layout';
 import WCClientInfo from './components/WCClientInfo';
 import useWalletConnect from './hooks/useWalletConnect';
-
-const StyledTitle = styled(Title)`
```
Comment (DaniSomoza) [inline apps/wallet-connect/src/App.tsx:13]: I moved all the styling things in the bottom of this file

File: apps/wallet-connect/src/App.tsx [hunk 2]
```diff
@@ -197,14 +139,14 @@ const App = () => {
     <>
       {/* WalletConnect */}
       {wcClientData && (
-        <StyledParagraph size="sm">
+        <StyledWrapper>
```
Comment (DaniSomoza) [inline apps/wallet-connect/src/App.tsx:142]: This change is to remove the warning in the console:
Captura de pantalla 2022-01-18 a las 17 47 59

File: apps/wallet-connect/src/components/WCClientInfo.tsx [hunk 3]
```diff
@@ -26,7 +26,7 @@ type Props = {
 const WCClientInfo = ({ name, url, iconSrc }: Props) => {
   return (
     <Wrapper>
-      <Image src={iconSrc} />
+      <Image src={iconSrc} role="img" />
```
Comment (DaniSomoza) [inline apps/wallet-connect/src/components/WCClientInfo.tsx:29]: This is for Accessibility purposes, `ARIA img role` is used to identify this div element as an image.

see docs for more info: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/img_role

File: package.json [hunk 4]
```diff
@@ -9,7 +9,8 @@
     "start:tx-builder": "yarn workspace tx-builder start",
     "start:walletconnect": "yarn workspace wallet-connect start",
     "start:drain-safe": "yarn workspace drain-safe start",
-    "test": "yarn workspaces run test --watchAll=false"
+    "test": "yarn workspaces run test --watchAll=false",
+    "test:coverage": "yarn workspaces run test --coverage --watchAll=false"
```
Comment (DaniSomoza) [inline package.json:13]: Added a `coverage` script to `package.json`

**Review by mikheevm — APPROVED** [review]
muy buen trabajo

**Comment by JagoFigueroa** [comment]
Talked with Dani privately about a couple of issues / improvements I see that could be performed for current implementation. Back to mi hibernation for a bit ;)

**Comment by mikheevm** [comment]
> Talked with Dani privately about a couple of issues / improvements I see that could be performed for current implementation. Back to my hibernation for a lil bit ;)

Would you please post them publicly? I think it's better to keep them public, so everyone is aware

**Comment by JagoFigueroa** [comment]
Sure thing and sorry, did so as a convenience due to my current state.

Here is the list: 

- I can see that once the scan is started after a connection is performed, the cam keeps recording until the app is quitted. I would just limit the cam to work only when scanning pop up is present. As it is an app that on paper will be in the foreground more often than not, if the user were to connect + immediately use the app it looks like we are spying or something as the cam will be on.

Minor stuff / future enhancements

- I think it would be nice to add something like this to provide some feedback if the user wants to quit the scanning process. (esc key is doing the trick as of now)

Screenshot 2022-01-21 at 09 37 14
(working from bed swag)

When WC input is focused, remove the grey text as to match address input component styling
Screenshot 2022-01-21 at 07 25 59

**Comment by mikheevm** [comment]
> Sure thing and sorry, did so as a convenience due to my current state.
> 
> 
> 
> Here is the list: 
> 
> 
> 
> - I can see that once the scan is started, the cam keeps recording until the app is quitted. I would just limit the cam to work only when scanning pop up is present. As it is an app that on paper will be in the foreground more often than not, if the user were to connect + immediately use the app it looks like we are spying or something as the cam will be on.
> 
> 
> 
> Minor stuff / future enhancements
> 
> 
> 
> - I think it would be nice to add something like this to provide some feedback if the user wants to quit the scanning process. (esc key is doing the trick as of now)
> 
> 
> 
> (working from bed swag)
> 
> Screenshot 2022-01-21 at 09 37 14
> 
> 
> 
> 
> 
> - When WC input is focused, remove the grey text as to match address input component styling
> 
> 
> 
> Screenshot 2022-01-21 at 07 25 59
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

I am sorry ser 🙁, hope you get better soonn 🤞

**Comment by DaniSomoza** [comment]
Improvements Addressed!

Captura de pantalla 2022-01-21 a las 12 49 17

I added an extra test to make sure that we can close the Dialog and the camera connection is closed:

```
    it('Closes webcam connection by closing the dialog', async () => {
      const stopWebcamSpy = jest.fn();

      const openWebcamSpy = jest.fn().mockImplementation(() => ({
        getTracks: () => [
          {
            stop: stopWebcamSpy,
          },
        ],
      }));

      Object.defineProperty(window.navigator, 'mediaDevices', {
        writable: true,
        value: {
          getUserMedia: openWebcamSpy,
        },
      });

      Object.defineProperty(window.HTMLCanvasElement.prototype, 'getContext', {
        writable: false,
        value: () => {
          return {
            drawImage: jest.fn(),
            getImageData: jest.fn(),
          };
        },
      });

      renderWithProviders(<App />);

      expect(openWebcamSpy).not.toHaveBeenCalled();
      expect(stopWebcamSpy).not.toHaveBeenCalled();

      const openDialogNode = screen.getByRole('button');

      // we open webcam dialog
      fireEvent.click(openDialogNode);

      const scanQRCodeDialog = await screen.findByRole('dialog');

      expect(scanQRCodeDialog).toBeInTheDocument();

      // only webcam connection should be called at this point of the test
      expect(openWebcamSpy).toHaveBeenCalled();
      expect(stopWebcamSpy).not.toHaveBeenCalled();

      // we close webcam dialog
      const closeQRCodeDialogButton = screen.getByLabelText('Close scan QR code dialog');

      expect(closeQRCodeDialogButton).toBeInTheDocument();
      fireEvent.click(closeQRCodeDialogButton);

      await waitForElementToBeRemoved(() => screen.queryByRole('dialog'));

      // webcam connection should be closed
      expect(stopWebcamSpy).toHaveBeenCalled();
```

CC: @JagoFigueroa

**Comment by JagoFigueroa** [comment]
Working like a charm now, thank you for the fixes! ;)

**Pull request merged**